### DISCUSSION TITLE
Firedrake not installing due to issues with PETSC dependencies.Dealing with one dependency causes the others to fail again.

### DISCUSSION THREAD

**Discussion by TehmoorG** [firedrakeproject/firedrake#3610] [Firedrake support]
[configure.log](https://github.com/user-attachments/files/15550731/configure.log)
[firedrake-install.log](https://github.com/user-attachments/files/15550733/firedrake-install.log)

We are encountering multiple issues during the installation of Firedrake on a MacBook. The primary challenges include failed downloads and extraction errors for various PETSc dependencies like HWLOC, MPICH, and HDF5. These issues appear to stem from corrupted or incomplete downloads, incorrect file paths, and potential conflicts with existing system configurations. Additionally, setting the correct environment variables for PETSc directories and architectures has proven difficult, leading to repeated installation failures despite following the specified procedures.

When I have with the HWLOC installation and then tried dealing with MPICH the HWLOC has an issue again. I have a link of a chatgpt chat showing the things I have attempted:
https://chatgpt.com/share/609f78ed-9249-46a2-8b8b-5f9065910851

I have tried but can't seem to fix the issue and I am quite new to programming so finding this overwhelming. Thanks for any help

**Comment by TehmoorG** [comment]
I also wanted to give the latest error message on the file:

*********************************************************************************************
           UNABLE to CONFIGURE with GIVEN OPTIONS (see configure.log for details):
---------------------------------------------------------------------------------------------
  Error during download/extract/detection of HWLOC:
  Unable to download package HWLOC from:
  https://download.open-mpi.org/release/hwloc/v2.10/hwloc-2.10.0.tar.gz
  * If URL specified manually - perhaps there is a typo?
  * If your network is disconnected - please reconnect and rerun ./configure
  * Or perhaps you have a firewall blocking the download
  * You can run with --with-packages-download-dir=/adirectory and ./configure will instruct
  you what packages to download manually
  * or you can download the above URL manually, to
  /yourselectedlocation/hwloc-2.10.0.tar.gz
  and use the configure option:
  --download-hwloc=/yourselectedlocation/hwloc-2.10.0.tar.gz
  Unable to download package HWLOC from:
  https://web.cels.anl.gov/projects/petsc/download/externalpackages/hwloc-2.10.0.tar.gz
  * If URL specified manually - perhaps there is a typo?
  * If your network is disconnected - please reconnect and rerun ./configure
  * Or perhaps you have a firewall blocking the download
  * You can run with --with-packages-download-dir=/adirectory and ./configure will instruct
  you what packages to download manually
  * or you can download the above URL manually, to
  /yourselectedlocation/hwloc-2.10.0.tar.gz
  and use the configure option:
  --download-hwloc=/yourselectedlocation/hwloc-2.10.0.tar.gz
*********************************************************************************************


2024-06-03 18:42:10,471 DEBUG  Popping path '/Users/fargenakazi/Library/CloudStorage/OneDrive-ImperialCollegeLondon/firedrake/src/petsc'
2024-06-03 18:42:10,471 DEBUG  New path '/Users/fargenakazi/Library/CloudStorage/OneDrive-ImperialCollegeLondon/firedrake/src'
2024-06-03 18:42:10,480 INFO   
Install log saved in /Users/fargenakazi/Library/CloudStorage/OneDrive-ImperialCollegeLondon/firedrake-install.log

**Reply by connorjward** [reply]
The information you got from ChatGPT is not correct. You should not need to set `PETSC_CONFIGURE_OPTIONS` at all for a regular Firedrake installation.

From your install log I see three immediate issues (beyond this):
- You are using the wrong Python. You should be using the Python installed via Homebrew (i.e. run `$(brew --prefix)/bin/python3 firedrake-install`).
- You seem to have quite an old Xcode/compiler installed. I recommend updating Xcode to the latest version (and then also running `brew upgrade`) before trying again.
- You should not try to install Firedrake into a cloud drive (like OneDrive). I think that that is very unlikely to work.

**Reply by TehmoorG** [reply]
Hi, Ive tried to do as you asked. I first installed the latest version of Xcode from this website:
https://xcodereleases.com/ and then I clicked on the downloaded app. I then ran `brew upgrade` in my terminal. I did this on my computer and not one drive. I tried to run firedrake install with the command you suggested but this didn't work so I used the following:


```
fargenakazi@Fargenas-MBP fire % (brew --prefix)/bin/python3 firedrake-install)
zsh: parse error near `/bin/python3'
fargenakazi@Fargenas-MBP fire % which python3
/Library/Frameworks/Python.framework/Versions/3.11/bin/python3
fargenakazi@Fargenas-MBP fire % ls
firedrake-install
fargenakazi@Fargenas-MBP fire % brew --prefix
/usr/local


fargenakazi@Fargenas-MBP fire % /Library/Frameworks/Python.framework/Versions/3.11/bin/python3 firedrake-install
Running firedrake-install
Installing command line tools...
Checking brew doctor...
```


[configure.log](https://github.com/user-attachments/files/15567350/configure.log)
[firedrake-install.log](https://github.com/user-attachments/files/15567351/firedrake-install.log)

**Reply by connorjward** [reply]
You still have the same old compiler. Are you sure that you correctly installed the latest Xcode? I think you can check this if you run
```
$ xcodebuild -version
```
and I think the current latest is 15.3.

Also I think I could have been clearer telling you the right way to get the right Python to use. You want to run:
```
$ $(brew --prefix)/bin/python3 firedrake-install
```
One of those `$` signs means "run this at the terminal" and the other means "expand `brew --prefix`". You can skip all the confusion by just running, in your case:
```
$ /usr/local/bin/python3 firedrake-install
```

**Reply by connorjward** [reply]
Also, you appear to just be failing to download an archive. Do you have a stable internet connection?

**Reply by TehmoorG** [reply]
Hi hope you are well,
I do have a stable internet connection and I have done exactly as you asked. Here is the terminal showing this
```
fargenakazi@Fargenas-MBP fire % xcodebuild -version                           
Xcode 15.4
Build version 15F31d
fargenakazi@Fargenas-MBP fire % brew upgrade
==> Downloading https://formulae.brew.sh/api/formula.jws.json
############################################################################################################################################################################## 100.0%
fargenakazi@Fargenas-MBP fire % /usr/local/bin/python3 firedrake-install                                
Traceback (most recent call last):
  File "/Users/fargenakazi/fire/firedrake-install", line 20, in <module>
    from pkg_resources.extern.packaging.version import Version, InvalidVersion
ModuleNotFoundError: No module named 'pkg_resources'


During handling of the above exception, another exception occurred:


Traceback (most recent call last):
  File "/Users/fargenakazi/fire/firedrake-install", line 22, in <module>
    from packaging.version import Version, InvalidVersion
ModuleNotFoundError: No module named 'packaging'

fargenakazi@Fargenas-MBP fire % ls
firedrake-install

fargenakazi@Fargenas-MBP fire % /usr/local/bin/python3 --version                                        
Python 3.12.3
fargenakazi@Fargenas-MBP fire % which python3                                                           
/Library/Frameworks/Python.framework/Versions/3.11/bin/python3



fargenakazi@Fargenas-MBP fire % /Library/Frameworks/Python.framework/Versions/3.11/bin/python3 firedrake-install
Running firedrake-install
Installing command line tools...
Checking brew doctor...
Installing required packages via homebrew...
Required package 'gcc' is already installed via Homebrew.
Required package 'autoconf' is already installed via Homebrew.
Required package 'make' is already installed via Homebrew.
Required package 'automake' is already installed via Homebrew.
Required package 'cmake' is already installed via Homebrew.
Required package 'libtool' is already installed via Homebrew.
Required package 'boost' is already installed via Homebrew.
Creating firedrake venv in '/Users/fargenakazi/fire/firedrake'.
BLAS config is {'BLAS': '/Users/fargenakazi/fire/firedrake/src/petsc/default', 'OPENBLAS': '/Users/fargenakazi/fire/firedrake/src/petsc/default', 'LDFLAGS': '-L/usr/local/opt/gcc/lib/gcc/14'}
Configuration saved to /Users/fargenakazi/fire/firedrake/.configuration.json
Cloning firedrake
```

The command prompt you suggested using to run the install didn't work Im not sure why but I ran the brew python version 3.11 instead. The install still hasn't worked and I have attached the files again below. Thank you for all your help.
[configure.log](https://github.com/user-attachments/files/15687476/configure.log)
[firedrake-install.log](https://github.com/user-attachments/files/15687477/firedrake-install.log)

**Reply by Ig-dolci** [reply]
I think you still are not installing Firedrake with homebrew Python. To ensure you are installing Firedrake with Homebrew's Python:

1. **Find Homebrew Path**:
   Run `brew --prefix` in your terminal to get the Homebrew installation path. For example,  in my laptop the command `brew --prefix` returns  `/opt/homebrew`.

2. **Install Firedrake**:
   Use the full path to Homebrew's Python to install Firedrake. Again using my laptop command as example:
   ```
   /opt/homebrew/bin/python3 -m firedrake-install
   ```
but the command `$(brew --prefix)/bin/python3 firedrake-install` is still better.

**Comment by JDBetteridge** [comment]
@TehmoorG if you like the idea of using an interface similar to ChatGPT, we have an active slack channel, which you can sign up to here: https://www.firedrakeproject.org/contact.html#slack

The only difference is that the responses might actually be correct!

**Reply by TehmoorG** [reply]
Will do :)
